File: gunhoflash/product-of-array-except-self.py
from typing import List

class Solution:
  def productExceptSelf(self, nums: List[int]) -> List[int]:
    lenNums = len(nums)
    answer = [1] * lenNums

    serialProduct = 1
    for i in range(lenNums):
      answer[i] *= serialProduct
      serialProduct *= nums[i]

    serialProduct = 1
    for i in range(lenNums - 1, -1, -1):
      answer[i] *= serialProduct
      serialProduct *= nums[i]

    return answer

File: gunhoflash/product-of-array-except-self.ts
function productExceptSelf(nums: number[]): number[] {
  const answer = Array<number>(nums.length).fill(1);

  const reduceFunction = (serialProduct: number, num: number, i: number) => {
    answer[i] *= serialProduct;
    return serialProduct * num;
  };

  nums.reduce(reduceFunction, 1);
  nums.reduceRight(reduceFunction, 1);

  return answer;
}
